launchable/commands/record/commit.py

```diff
@@ -10,11 +19,67 @@
               default=os.getcwd(),
               type=str
               )
-def commit(source):
-    token, org, workspace = parse_token()
+@click.option('--executable',
+              help="collect commits with Jar or Docker",
+              type=click.Choice(['jar', 'docker']),
+              default='jar'
+              )
+def commit(source, executable):
     source = os.path.abspath(source)
+
+    if executable == 'jar':
+        download_jar()
+        exec_jar(source)
+    elif executable == 'docker':
+        exec_docker(source)
+
+
+def download_jar():
+    pathlib.Path(jar_dir_path).mkdir(parents=True, exist_ok=True)
+
+    res = requests.get(ingester_jar_url, stream=True)
+    total_length = int(res.headers.get('content-length'))
```

Comment:
I'm also mildly concerned that every invocation of build recording now creates a request to S3 and any problems there will render the entire command inoperable. We already know for example that BMW's build environment needs IP-based whitelisting, so I think this would create an extra step for them.

If we download a file into a temporary file, then rename it to the expected name, then `os.path.exists(...)` would be sufficient to verify that completeness of a file, so that would work avoid this problem. In this way, BMW can also just manually download a file and place it in the location they expect. Also handy for people who create a builder image.

Reply:
Thank you for your suggestion. I implemented moving `jar` from the temporary directory to the cache directory. As a result, the CLI doesn't need to access S3 if the jar file was downloaded.